esp-config/src/generate.rs

```diff
@@ -1,77 +1,246 @@
-                    _ => usize::from_str_radix(&s, 10),
+            },
+            Value::Integer(_) => {
+                let inner = match s.as_bytes() {
+                    [b'0', b'x', ..] => i128::from_str_radix(&s[2..], 16),
+                    [b'0', b'o', ..] => i128::from_str_radix(&s[2..], 8),
+                    [b'0', b'b', ..] => i128::from_str_radix(&s[2..], 2),
+                    _ => i128::from_str_radix(&s, 10),
                 }
-                .map_err(|_| ParseError(format!("Invalid unsigned numerical value: {}", s)))?,
-            ),
-            Value::String(_) => Value::String(String::from(s)),
+                .map_err(|_| Error::parse(format!("Expected valid intger value, found: '{s}'")))?;
+
+                Value::Integer(inner)
+            }
+            Value::String(_) => Value::String(s.into()),
         };
+
         Ok(())
     }
 
-    fn as_string(&self) -> String {
+    /// Convert the value to a [bool].
+    pub fn as_bool(&self) -> bool {
+        match self {
+            Value::Bool(value) => *value,
+            _ => panic!("attempted to convert non-bool value to a bool"),
+        }
+    }
+
+    /// Convert the value to an [i128].
+    pub fn as_integer(&self) -> i128 {
+        match self {
+            Value::Integer(value) => *value,
+            _ => panic!("attempted to convert non-integer value to an integer"),
+        }
+    }
+
+    /// Convert the value to a [String].
+    pub fn as_string(&self) -> String {
+        match self {
+            Value::String(value) => value.to_owned(),
+            _ => panic!("attempted to convert non-string value to a string"),
+        }
+    }
+
+    /// Is the value a bool?
+    pub fn is_bool(&self) -> bool {
+        matches!(self, Value::Bool(_))
+    }
+
+    /// Is the value an integer?
+    pub fn is_integer(&self) -> bool {
+        matches!(self, Value::Integer(_))
+    }
+
+    /// Is the value a string?
+    pub fn is_string(&self) -> bool {
+        matches!(self, Value::String(_))
+    }
+}
+
+impl fmt::Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Value::Bool(b) => write!(f, "{b}"),
+            Value::Integer(i) => write!(f, "{i}"),
+            Value::String(s) => write!(f, "{s}"),
+        }
+    }
+}
+
+/// Configuration value validation functions.
+pub enum Validator {
+    /// Only allow negative integers, i.e. any values less than 0.
+    NegativeInteger,
+    /// Only allow non-negative integers, i.e. any values greater than or equal
+    /// to 0.
+    NonNegativeInteger,
+    /// Only allow positive integers, i.e. any values greater than to 0.
+    PositiveInteger,
+    /// Ensure that an integer value falls within the specified range.
+    IntegerInRange(Range<i128>),
+    /// A custom validation function to run against any supported value type.
+    Custom(Box<dyn Fn(&Value) -> Result<(), Error>>),
```

Comment:
This is more of a general question than a review comment. Can we somehow cross-validate values using `Custom`? I can't figure out how right now, but I think it would be useful if we could - there are value combinations in esp-wifi that don't make much sense, for example.

Reply:
Is this in the scope of esp-config, or should this be managed in the crate itself? The downstream crate will have the config in a const context, it can do the required checks there to ensure that if X is enabled Y also is or isn't etc.

Reply:
> The downstream crate will have the config in a const context, it can do the required checks there to ensure that if X is enabled Y also is or isn't etc.

That would separate the declaration of the variables (along with their individual validation rules, which are in build.rs) from the cross-validation rules which would be in `lib.rs`. I'm not sure if that is the ideal state we want to have.

Reply:
That's true, but we can actually couple the cross-validations into build.rs too (something I only just remembered), as `generate_config` returns a hashmap of the key/set values. 

I guess what I'm saying is that cross-validaton _may_ come to esp-config in the future, but I think its out of scope of individual config validation, i.e let's not try and abuse `Custom` to do it.

Reply:
I'd say anything beyond validating a single value is out of scope - it can be arbitrary complex and probably easier to handle by custom code in `build.rs` than trying to come up with something smart